aspnetcore/signalr/hub-filters.md

```diff
@@ -0,0 +1,193 @@
+        options.AddFilter<CustomFilter>();
+    }).AddHubOptions<MyHub>(options =>
+    {
+        // Local filters will run second
+        options.AddFilter<CustomFilter2>();
+    });
+}
+```
+
+There are 3 ways to add hub filters.
+1. Add a filter by concrete type.
+
+```csharp
+hubOptions.AddFilter<TFilter>();
+```
+
+This will be resolved from DI or type activated.
+
+2. Add a filter by runtime type.
+
+```csharp
+hubOptions.AddFilter(typeof(TFilter));
+```
+
+This will be resolved from DI or type activated.
+
+3. Add a filter by instance.
+
+```csharp
+hubOptions.AddFilter(new MyFilter());
+```
+
+This instance will be used like a singleton, all hub method invocations will use the same instance.
+
+Hub filters are created and disposed per hub invocation. If you want to store global state in the filter, or no state, then we recommend adding the hub filter type to DI as a singleton for better performance or add the filter as an instance if you can.
+
+## Create hub filters
+
+Create a filter by declaring a class that inherits from `IHubFilter`, and add the `InvokeMethodAsync` method. There is also `OnConnectedAsync` and `OnDisconnectedAsync` that can optionally be implemented to wrap the `OnConnectedAsync` and `OnDisconnectedAsync` hub methods respectively.
+
+```csharp
+public class CustomFilter : IHubFilter
+{
+    public async ValueTask<object> InvokeMethodAsync(HubInvocationContext invocationContext, Func<HubInvocationContext, ValueTask<object>> next)
+    {
+        Console.WriteLine($"Calling hub method '{invocationContext.HubMethodName}'");
+        try
+        {
+            return await next(invocationContext);
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine($"Exception calling '{invocationContext.HubMethodName}'");
+            throw ex;
+        }
+    }
+
+    // Optional method
+    public Task OnConnectedAsync(HubLifetimeContext context, Func<HubLifetimeContext, Task> next)
+    {
+        return next(context);
+    }
+
+    // Optional method
+    public Task OnDisconnectedAsync(HubLifetimeContext context, Exception exception, Func<HubLifetimeContext, Exception, Task> next)
+    {
+        return next(context, exception);
+    }
+}
+```
+
+Filters are very similar to middleware, `next` will invoke the next filter and the final filter will invoke the hub method. Filters can also store the result from awaiting `next` and run logic after the hub method has been called before returning the result from `next`.
+
+If a filter wants to skip a hub method invocation we recommend throwing a `HubException` and not calling `next` so the client will receive an error if it was expecting a result.
+
+## Using hub filters
+
+When writing the filter logic, try to make it generic by using attributes on hub methods instead of checking for hub method names.
+
+For example, lets write a filter that will check an argument to a hub method for banned phrases and replace any phrases it finds with "***".
```

Comment:
The triple asterisk should be code-fenced to avoid confusion with Markdown formatting syntax.